fix(terminal): reset `w_leftcol` after resizing terminal

def3r opened the pull request:
Problem: windows may scroll horizontally upon resize using the old terminal
size, which may be unnecessary and cause the content to be partially out-of-view.

Solution: reset the horizontal scroll after resizing.

Closes #35331

seandewar reviewed (commented):
Doing this prevents horizontal scrolling from working even in Normal mode, and can cause the cursor to draw outside the viewport of the window it's in.

```vim
set splitright splitbelow
call chansend(nvim_open_term(0, #{}), "a"->repeat(99999))
normal! G
split
vnew
vsplit
wincmd 2W
" Keep moving the cursor to the right until it leaves the bottom-left window's viewport
call timer_start(20, {->execute('normal! l')}, #{repeat: -1})
```

def3r commented:
I came up with this solution for now. We take a flag `may_resize` and set it in `terminal_check_size`. This flag is supposed to tell whether the proc running inside the terminal has responded to the resize. We set w_leftcol to 0 if may_resize is set. 

This seems to work but probably isn't the right idea. I feel terminal_recieve should have a better way to indicate that resize has happened.

seandewar commented:
Yeah I think that's better. It's simple and resizing shouldn't occur often enough for performance to be a concern. (plus `curs_columns` isn't always called for non-current windows, so a loop may have been needed anyway)

Be sure to add some tests, and maybe a comment above the loop like:
```c
// Resized window may have scrolled horizontally to keep its cursor in-view using the terminal's old
// size. Reset the scroll, and let curs_columns correct it if that sends the cursor out-of-view.
```

def3r committed:
fix(terminal): reset `w_leftcol` after resizing terminal

Problem: windows may scroll horizontally upon resize using the old terminal
size, which may be unnecessary and cause the content to be partially out-of-view.

Solution: reset the horizontal scroll after resizing.

seandewar commented on test/functional/terminal/window_spec.lua:
```diff
@@ -33,6 +34,48 @@ describe(':terminal window', function()
+    command('vertical resize 15')
+    command('redraw!')
+    screen:expect([[
+      nteract $ AAAA^ │                                  |
```
This isn't right. `curs_columns` needs to happen after the cursor is adjusted.

seandewar reviewed (commented):
I'll try some fixups in a bit.

seandewar reviewed (approved):
Thanks, your comment saved me a bunch of time! 😅

I think this is fine now. Might exist other cases where a scroll reset could be useful; may look into that at some point.
